pkg/trtl/backup.go

```diff
@@ -0,0 +1,190 @@
+package trtl
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/rotationalio/honu"
+	honuldb "github.com/rotationalio/honu/engines/leveldb"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/syndtr/goleveldb/leveldb"
+	ldbstore "github.com/trisacrypto/directory/pkg/gds/store/leveldb"
+	"github.com/trisacrypto/directory/pkg/trtl/config"
+	"github.com/trisacrypto/directory/pkg/utils"
+)
+
+// BackupManager runs as an independent service which periodically copies the trtl
+// storage to a compressed backup location on disk.
+type BackupManager struct {
+	conf config.Config
+	db   *honu.DB
+	stop chan struct{}
+}
+
+func NewBackupManager(s *Server) (*BackupManager, error) {
+	return &BackupManager{
+		conf: s.conf,
+		db:   s.db,
+		stop: make(chan struct{}),
+	}, nil
+}
+
+// Runs the main BackupManager routine which periodically wakes up and creates a backup
+// of the trtl database.
+func (m *BackupManager) Run() {
+	backupDir, err := m.getBackupStorage()
+	if err != nil {
+		log.Fatal().Err(err).Msg("trtl backup manager cannot access backup directory")
+	}
+
+	ticker := time.NewTicker(m.conf.Backup.Interval)
+	log.Info().Dur("interval", m.conf.Backup.Interval).Str("store", backupDir).Msg("trtl backup manager started")
+
+backups:
+	for {
+		// Wait for next tick
+		<-ticker.C
+
+		// Begin the backup process
+		start := time.Now()
+		log.Debug().Msg("starting backup of trtl database")
+
+		// Perform the backup
+		if err = m.backup(backupDir); err != nil {
+			// Do not continue if there was a backup error; all code in the rest of the
+			// loop should expect that the backup was successful.
+			// NOTE: using WithLevel and Fatal does not Exit the program like log.Fatal()
+			// this ensures that we issue a CRITICAL severity without stopping the server.
+			log.WithLevel(zerolog.FatalLevel).Err(err).Msg("could not backup database")
+			continue backups
+		} else {
+			log.Info().Dur("duration", time.Since(start)).Msg("trtl backup complete")
+		}
+
+		// Remove any previous backups that may be in the directory
+		// NOTE: this requires the backup to write filenames as trtldb-200601021504.*
+		var archives []string
+		if archives, err = listArchives(backupDir); err != nil {
+			log.Error().Err(err).Msg("could not list backup directory")
+		} else {
+			if len(archives) > m.conf.Backup.Keep {
+				var removed int
+				for _, archive := range archives[:len(archives)-m.conf.Backup.Keep] {
+					log.Debug().Str("archive", archive).Msg("deleting archive")
+					if err = os.Remove(archive); err == nil {
+						removed++
+					}
+				}
+				log.Debug().Int("kept", m.conf.Backup.Keep).Int("removed", removed).Msg("backup directory cleaned up")
+			}
+		}
+
+		select {
+		case <-m.stop:
+			log.Warn().Msg("trtl backup manager received stop signal")
+			return
+		default:
+		}
+	}
+}
+
+func (m *BackupManager) Shutdown() error {
+	if m.conf.Backup.Enabled && m.stop != nil {
+		close(m.stop)
```

Comment:
When we call `Shutdown` shouldn't we send a stop signal rather than close the channel? I'm sorry, I'm just not remembering what we intended for this - was it primarily to stop tests?